### PR TITLE
Initial defaults work

### PR DESCRIPTION
**Pull request by edwardpeters**
This PR implements some "Default" generation to make Data out of Concepts, or to fill in limited Data (such as a partial record) to meet a Concept.

### CODE REVIEW

File: morphir/src/org/finos/morphir/datamodel/Defaults.scala [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+import zio.*
+
+type DefaultTask = Task[Data]
+
+sealed trait MDMDefaultError                        extends Throwable
+case class Unimplemented(msg: String)               extends MDMDefaultError
+case class NoDefaultNothing()                       extends MDMDefaultError
+case class Unfillable(data: Data, concept: Concept) extends MDMDefaultError
+
+trait Defaults {
```
Comment (DamianReeves) [inline morphir/src/org/finos/morphir/datamodel/Defaults.scala:11]: I'd do this as a typeclass instead of an interface

File: morphir/src/org/finos/morphir/datamodel/Defaults.scala [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  def defaultFloat: DefaultTask = ZIO.succeed(Data.Float(0))
+  def defaultLocalTime: DefaultTask =
+    ZIO.succeed(Data.LocalTime(java.time.LocalTime.MIDNIGHT))
+  def defaultInt32: DefaultTask  = ZIO.succeed(Data.Int32(0))
+  def defaultByte: DefaultTask   = ZIO.succeed(Data.Byte(0))
+  def defaultUnit: DefaultTask   = ZIO.succeed(Data.Unit)
+  def defaultString: DefaultTask = ZIO.succeed(Data.String(""))
+  def defaultOrder: DefaultTask  = ZIO.succeed(Data.Order.apply(0))
+  def defaultDayOfWeek: DefaultTask =
+    ZIO.succeed(Data.DayOfWeek(java.time.DayOfWeek.of(0)))
```
Comment (DamianReeves) [inline morphir/src/org/finos/morphir/datamodel/Defaults.scala:83]: Why do we need ZIO for this?

Comment (DamianReeves) [reply morphir/src/org/finos/morphir/datamodel/Defaults.scala:83]: Was your intention to use ZIO's `Gen` to provide generators?

Comment (edwardpeters) [reply morphir/src/org/finos/morphir/datamodel/Defaults.scala:83]: I used ZIO for a few reasons:

- Some of these can already fail (Concept.Nothing should not be able to make a default)
- Even if this implementation always succeeds, users might want to specify failures for some cases - for instance, not creating defaults for some privileged types - so I wanted the interface to allow for such
- Future-proofing (we recently had issues with contracts/exceptions because a function we didn't originally think could fail actually could)
- Consistency with ZIO usage thru-out the project (basically that's just why I used ZIO rather than Either)
-

Comment (edwardpeters) [reply morphir/src/org/finos/morphir/datamodel/Defaults.scala:83]: I had no plans regarding ZIO's `Gen`

Comment (michelchan) [reply morphir/src/org/finos/morphir/datamodel/Defaults.scala:83]: Some thoughts that popped up in my head from the reply:
1) The exception would be the default and would be the same exception either way. It's not a side effect that would require ZIO.
2) Users specifying specific failures defeats the purpose of a default. If users want special logic, they shouldn't be doing that here.
3) Defaults should never side effect.

File: morphir/src/org/finos/morphir/datamodel/Defaults.scala (hunk 1 above)
Comment (michelchan) [reply morphir/src/org/finos/morphir/datamodel/Defaults.scala:11]: Considering these are just defaults to existing Concepts, I'd do them in their corresponding case object instead of an extra interface.

Comment (edwardpeters) [reply morphir/src/org/finos/morphir/datamodel/Defaults.scala:11]: My only problem with that is that they're recursively defined - users may want to use the `defaultRecord`'s behavior for **how** a record should be filled with default values, but need to provide their own definition for what the default of a particular type is. Resolving that - so that `morphir-scala`'s `defaultRecord` code will make recursive calls to a user-defined `defaultLocalDate` function - is the reason I structured it this way.

File: morphir/src/org/finos/morphir/datamodel/Defaults.scala (hunk 2 above)
Comment (edwardpeters) [reply morphir/src/org/finos/morphir/datamodel/Defaults.scala:83]: I'm not clear what you mean by 1, can you rephrase?

More generally, I feel like I may have stepped in it by naming this "Default" - the behavior of filling in a `Concept` with some data is something that can call for specific implementations and specific failures. Would you be happier if this were just not named Default?

Comment (edwardpeters) [reply morphir/src/org/finos/morphir/datamodel/Defaults.scala:83]: Oh, one specific thing - if these aren't allowed to fail, what should the default for `Concept.Nothing` be?

File: morphir/src/org/finos/morphir/datamodel/Concept.scala [hunk 3]
```diff
@@ -229,4 +234,115 @@ object Concept {
...
+      default(concept.okType).map(Data.Result.Ok(_, concept))
+    def defaultSet(concept: Concept.Set): DefaultEither = Right(Data.Set.empty(concept.elementType))
+    def defaultTuple(concept: Concept.Tuple): DefaultEither = for {
+      elems <- collectAll(concept.values, default)
+      res = Data.Tuple(elems)
+    } yield res
+  }
+
+  object Defaults {
+    type DefaultEither = Either[MDMDefaultError, Data]
```
Comment (DamianReeves) [inline morphir/src/org/finos/morphir/datamodel/Concept.scala:328]: Can we call this `Result` and make it an `opaque type`:

```scala
opaque type Result <: Either[MDMDefaultError, Data] = Either[MDMDefaultError, Data]
```

As of late I am trying to standardize in many of these things in a way inspired by how Rust does this. And I like using the Companion object to "namespace" the type.

File: morphir/src/org/finos/morphir/datamodel/Concept.scala [hunk 4]
```diff
@@ -229,4 +234,115 @@ object Concept {
...
+    sealed trait MDMDefaultError  extends Throwable
```
Comment (DamianReeves) [inline morphir/src/org/finos/morphir/datamodel/Concept.scala:329]: In the same spirit of the Rust inspired naming lets just call this `Error`, it is already scoped by the `Defaults` companion object name.

File: morphir/src/org/finos/morphir/datamodel/Concept.scala [hunk 5]
```diff
@@ -229,4 +234,115 @@ object Concept {
...
+    case class NoDefaultNothing() extends MDMDefaultError
+
+    def collectAll[I, O, E](
```
Comment (DamianReeves) [inline morphir/src/org/finos/morphir/datamodel/Concept.scala:332]: Should this be publicly available? If it is an internal helper only make it `private` or package private via a `private[datamodel]` or `private[morphir]` access modifier.

File: morphir/src/org/finos/morphir/datamodel/Data.scala [hunk 6]
```diff
@@ -91,7 +91,7 @@ object Data {
    */
   case class Case(
       values: scala.List[(EnumLabel, Data)],
-      enumLabel: java.lang.String,
+      enumLabel: java.lang.String, // TODO: Why is this a string when the associated Concept is a label?
```
Comment (DamianReeves) [inline morphir/src/org/finos/morphir/datamodel/Data.scala:94]: Have you resolved or figured out this TODO? If so remove it.
